java/src/main/java/cucumber/runtime/java/JavaBackend.java

```diff
@@ -56,18 +57,25 @@ private Snippet createSnippet() {
 
     /**
      * The constructor called by reflection by default.
-     *
-     * @param resourceLoader
+     * 
+     * The given runtimeOptions may specify the class name of an ObjectFactory to use.
+     * An ObjectFactory can also be specified in cucumber.properties using the property name io.cucumber.core.backend.ObjectFactory
+     * or using the deprecated cucumber.api.java.ObjectFactory property.
      */
-    public JavaBackend(ResourceLoader resourceLoader, TypeRegistry typeRegistry) {
-        this(new ResourceLoaderClassFinder(resourceLoader, currentThread().getContextClassLoader()), typeRegistry);
+    public JavaBackend(ResourceLoader resourceLoader, TypeRegistry typeRegistry, RuntimeOptions runtimeOptions) {
+        this(new ResourceLoaderClassFinder(resourceLoader, currentThread().getContextClassLoader()), typeRegistry, runtimeOptions.getObjectFactory());
     }
 
-    private JavaBackend(ClassFinder classFinder, TypeRegistry typeRegistry) {
-        this(loadObjectFactory(classFinder, Env.INSTANCE.get(ObjectFactory.class.getName())), classFinder, typeRegistry);
+    private JavaBackend(ClassFinder classFinder, TypeRegistry typeRegistry, Class<? extends ObjectFactory> objectFactoryClass) {
+        this(loadObjectFactory(
+                classFinder,
+                (objectFactoryClass != null) ? objectFactoryClass.getName() : Env.INSTANCE.get(ObjectFactory.class.getName()),
+                Env.INSTANCE.get(cucumber.api.java.ObjectFactory.class.getName())), 
```

Comment:
Instead of using `cucumber.api.java.ObjectFactory.class.getName()` let's consider using the property `cucumber.object-factory` instead.